### PR TITLE
Project Plugins: Integrate plugins service.

### PR DESCRIPTION
**Pull request by luispadron**
Resolves https://github.com/tuist/tuist/issues/2021

### Short description 📝

This PR adds:

- The logic and integration of the Plugin service.
- Adds required code to generate projects using local & remote plugins.
- Adds a new fixture `app_with_plugin`
- Adds new feature test `plugins.feature`

This extends the code introduced in:

#2103
#2107
#2194  
#2095

### CODE REVIEW

File: Sources/TuistLoader/Loaders/CachedManifestLoader.swift [hunk 1]
```diff
@@ -1,6 +1,7 @@
 import Foundation
 import ProjectDescription
 import TSCBasic
+import TuistGraph
```
Comment (fortmarek) [inline Sources/TuistLoader/Loaders/CachedManifestLoader.swift:4]: alternatively, you could do `import struct TuistGraph.Plugins` if you did not want to prepend all the returned models with `ProjectDescription`. But the current solution is fine, too

File: Sources/TuistLoader/Loaders/ManifestLoader.swift [hunk 2]
```diff
@@ -245,7 +272,24 @@ public class ManifestLoader: ManifestLoading {
...
+            manifest == .config || manifest == .plugin,
+            command == "swiftc",
+            let errorMessage = String(data: standardError, encoding: .utf8) else { return }
+
+        let defaultHelpersName = ProjectDescriptionHelpersBuilder.defaultHelpersName
+        let helperModules = [defaultHelpersName] + plugins.projectDescriptionHelpers.map(\.name)
+        let hasImportedUnexpectedModule = helperModules.reduce(false) { $0 || errorMessage.contains($1) }
+        guard hasImportedUnexpectedModule else { return }
+
+        logger.error("Can't import \(defaultHelpersName) or plugins in \(manifest.fileName(path))")
```
Comment (fortmarek) [inline Sources/TuistLoader/Loaders/ManifestLoader.swift:293]: This is an interesting function - and it probably makes sense to optimize this as it can happen quite often and it might not be immediately obvious what went wrong from the Swift default message. But in that case I'd test this in `ManifestLoaderIntegrationTests`, so we ensure that the contract between the Swift error message and the custom logic for handling the error is not broken.

Apart from that, I'd maybe divide the error in the default helpers and plugins. I think it's really important to know which one is actually missing and provide a custom error message for plugins with some pointers what the user could to fix the issue - as well as with default helpers (eg we could write something like: "Please check that you have defined project description helpers in `Tuist/ProjectDescriptionHelpers`")

File: Sources/TuistLoader/ProjectDEscriptionHelpers/ProjectDescriptionHelpersBuilder.swift [hunk 3]
```diff
@@ -27,66 +33,130 @@ final class ProjectDescriptionHelpersBuilder: ProjectDescriptionHelpersBuilding
...
+    ) {
         self.projectDescriptionHelpersHasher = projectDescriptionHelpersHasher
         self.cacheDirectory = cacheDirectory
         self.helpersDirectoryLocator = helpersDirectoryLocator
     }
 
-    func build(at: AbsolutePath, projectDescriptionSearchPaths: ProjectDescriptionSearchPaths) throws -> AbsolutePath? {
-        guard let helpersDirectory = helpersDirectoryLocator.locate(at: at) else { return nil }
-        if let cachedPath = builtHelpers[helpersDirectory] { return cachedPath }
+    func build(
```
Comment (fortmarek) [inline Sources/TuistLoader/ProjectDEscriptionHelpers/ProjectDescriptionHelpersBuilder.swift:54]: maybe it'd be good to test the building logic for plugins, too? Either in a new unit test or in `ProjectDescriptionHelpersBuilderIntegrationTests`, what do you think?

File: features/plugins.feature [hunk 4]
```diff
@@ -0,0 +1,8 @@
+Feature: Generate a new project using Tuist with a plugin.
+
+  Scenario: The project is an iOS application with plugins (app_with_plugin)
```
Comment (fortmarek) [inline features/plugins.feature:3]: this is great 👏

**Review by fortmarek — COMMENTED** [review]
Looks great! The process of adding `plugins` everywhere is a bit tedious but it's probably the best way to do it (or at least I could not think of anything better 😐). Otherwise, I have just pointed out some nitpicks 🙂

File: Sources/ProjectDescription/PluginLocation.swift [hunk 5]
```diff
@@ -15,16 +15,6 @@ public struct PluginLocation: Codable, Equatable {
         PluginLocation(type: .local(path: path))
     }
 
-    /// A `URL` to a `git` repository pointing at a `branch`.
-    ///
-    /// Example:
-    /// ```
-    /// .git(url: "https://git/plugin.git", branch: "main")
-    /// ```
-    public static func git(url: String, branch: String) -> Self {
```
Comment (pepibumur) [inline Sources/ProjectDescription/PluginLocation.swift:24]: Thanks for removing this one @luispadron

**Review by kwridan — COMMENTED** [review]
Thanks for your work on this @luispadron 

I haven't followed the development of plugins closely so may have missed some prior context here - I have a suggestion to minimise the passing the plugins throughout the codebase _(sadly this will causes a few integration issues for us 😞 as we use `TuistGenerator` as a library that is independent of Tuist's Project Description / manifests etc...)_

If I followed the code, the reason why the plugins is needed to be passed around is so that the manifest loader can get a reference to them during the loading process?

Some things to consider:

- Something along the lines of `manifestLoader.add(plugin: Plugin)` within `Generator.swift` and this can internally store a plugin to use for subsequent loading calls
- Or perhaps another implementation of `ManifestLoader` that knows how to load plugins (such that its transparent to the rest of the pipeline?)

Hope this helps

**Review by pepibumur — APPROVED** [review]
Solid work @luispadron as always 👏 👏

**Comment by kwridan** [comment]
@luispadron here's a prototype of one of my suggestions:

https://github.com/tuist/tuist/tree/prototype/plugins-integration-1

It's a bit tricky overall, seeing Plugins loading requires Manifest loading and vice versa :D 

Hope this helps

File: Sources/TuistLoader/ProjectDEscriptionHelpers/ProjectDescriptionHelpersBuilder.swift (hunk 3 above)
Comment (luispadron) [reply Sources/TuistLoader/ProjectDEscriptionHelpers/ProjectDescriptionHelpersBuilder.swift:54]: Added!

File: Sources/TuistLoader/Loaders/ManifestLoader.swift (hunk 2 above)
Comment (luispadron) [reply Sources/TuistLoader/Loaders/ManifestLoader.swift]: I've split the logs and added another one for build errors, think I may need some help here in general though as there are some edge cases that we should handle to have the best logging we can. I think this can be its own ticket

File: Tests/TuistKitTests/ProjectEditor/ProjectEditorMapperTests.swift [hunk 6]
```diff
@@ -256,7 +256,7 @@ final class ProjectEditorMapperTests: TuistUnitTestCase {
         XCTAssertEqual(manifestOneTarget.platform, .macOS)
         XCTAssertEqual(manifestOneTarget.product, .staticFramework)
         XCTAssertEqual(manifestOneTarget.settings, expectedSettings(sourceRootPath: sourceRootPath))
-        XCTAssertEqual(manifestOneTarget.sources.map(\.path), [manifestPaths.last])
+        XCTAssertEqual(manifestOneTarget.sources.map(\.path), [manifestPaths.last!])
```
Comment (fortmarek) [inline Tests/TuistKitTests/ProjectEditor/ProjectEditorMapperTests.swift:259]: nitpick: I'd use `XCTUnwrap` instead of force unwrapping

File: Tests/TuistKitTests/ProjectEditor/ProjectEditorMapperTests.swift [hunk 7]
```diff
@@ -267,7 +267,7 @@ final class ProjectEditorMapperTests: TuistUnitTestCase {
         XCTAssertEqual(manifestTwoTarget.platform, .macOS)
         XCTAssertEqual(manifestTwoTarget.product, .staticFramework)
         XCTAssertEqual(manifestTwoTarget.settings, expectedSettings(sourceRootPath: sourceRootPath))
-        XCTAssertEqual(manifestTwoTarget.sources.map(\.path), [manifestPaths.first])
+        XCTAssertEqual(manifestTwoTarget.sources.map(\.path), [manifestPaths.first!])
```
Comment (fortmarek) [inline Tests/TuistKitTests/ProjectEditor/ProjectEditorMapperTests.swift:270]: I'd `XCTUnwrap` here, too

**Review by fortmarek — APPROVED** [review]
The registering of plugins made the code much cleaner, can't wait for this to land!
